### PR TITLE
local run UI

### PR DESCRIPTION
**Pull request by chkeita**
## Summary of the Pull Request

This PR add a UI to the local run. 
- The UI currently monitors the logs and  some of the directory created (the rest will be wired in a coming PR)
- pressing 'q' will quit the PR
- By default, the job directory is deleted when the ui quits unless the parameter 'keep_job_dir' is specified

### CODE REVIEW

**Comment by chkeita** [comment]
This is what is looks like so far.
I am currently fixing an issue when handling keyboard inputs
image

@bmc-msft  @ranweiler

File: src/agent/coverage/examples/block_coverage.rs [hunk 1]
```diff
@@ -55,7 +55,7 @@ fn main() -> Result<()> {
     };
 
     let mut cmd = Command::new(&opt.cmd[0]);
-    cmd.args(&opt.cmd[1..]);
+    cmd.stdin(Stdio::null()).args(&opt.cmd[1..]);
```
Comment (chkeita) [inline src/agent/coverage/examples/block_coverage.rs:58]: This is the fix for the issue i was having when reading inputs from the terminal.
It seems like at least one of the process we start with Command is reading from stdin.
So i added stdin(Stdio::null()) to all the command we launch

**Comment by bmc-msft** [comment]
Since this uses crossterm, what are your thoughts on using [IsTty](https://docs.rs/crossterm/0.19.0/crossterm/tty/trait.IsTty.html) and enabling it by default for the majority of tasks?

File: src/agent/onefuzz-agent/src/main.rs [hunk 2]
```diff
@@ -55,6 +53,7 @@ async fn run(args: ArgMatches<'_>) -> Result<()> {
 }
 
 fn licenses() -> Result<()> {
+    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info")).init();
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/main.rs:56]: IMO, we don't need to enable logging for this method.

File: src/agent/onefuzz-agent/src/local/common.rs [hunk 3]
```diff
@@ -197,7 +215,7 @@ pub fn get_synced_dir(
 // fuzzing tasks from generating random task id to using UUID::nil(). This
 // enables making the one-shot crash report generation, which isn't really a task,
 // consistent across multiple runs.
-pub fn build_common_config(args: &ArgMatches<'_>, generate_task_id: bool) -> Result<CommonConfig> {
+pub fn build_local_context(args: &ArgMatches<'_>, generate_task_id: bool) -> Result<LocalContext> {
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/common.rs:218]: Having a single context, does this mean that each task in the job shares the task_id?

**Review by bmc-msft — APPROVED** [review]
I think this is fine as is, and we can refine and update in follow-on PRs.

File: src/agent/onefuzz-agent/src/local/common.rs (hunk 3 above)
Comment (chkeita) [reply src/agent/onefuzz-agent/src/local/common.rs:210]: we manually generate a new task_id every time we pass the common_config to build a task specific config you can see it in libfuzzer.rs

File: src/agent/coverage/examples/block_coverage.rs [hunk 4]
```diff
@@ -63,7 +63,7 @@ fn main() -> Result<()> {
     let filter = opt.load_filter_or_default()?;
 
     let mut cmd = Command::new(&opt.cmd[0]);
-    cmd.args(&opt.cmd[1..]);
+    cmd.stdin(Stdio::null()).args(&opt.cmd[1..]);
```
[the remark this answers is not in the document]
Comment (ranweiler) [reply src/agent/coverage/examples/block_coverage.rs:66]: Nevermind, you did this!

File: src/agent/onefuzz-agent/src/tasks/fuzz/libfuzzer_fuzz.rs [hunk 5]
```diff
@@ -195,8 +195,13 @@ impl LibFuzzerFuzzTask {
             &self.config.common.setup_dir,
         );
         let mut running = fuzzer.fuzz(crash_dir.path(), local_inputs, &inputs)?;
+        let running_id = running.id();
 
-        let sys_info = task::spawn(report_fuzzer_sys_info(worker_id, run_id, running.id()));
+        let sys_info = task::spawn(async move {
+            report_fuzzer_sys_info(worker_id, run_id, running_id)
+                .await
+                .unwrap()
```
Comment (ranweiler) [inline src/agent/onefuzz-agent/src/tasks/fuzz/libfuzzer_fuzz.rs:203]: Why `unwrap()`?

File: src/agent/onefuzz-agent/src/local/cmd.rs [hunk 6]
```diff
@@ -70,8 +88,7 @@ pub fn args(name: &str) -> App<'static, 'static> {
             Arg::with_name(TIMEOUT)
                 .long(TIMEOUT)
                 .help("The maximum running time in seconds")
-                .takes_value(true)
-                .required(false),
+                .takes_value(true),
```
[the remark this answers is not in the document]
Comment (chkeita) [reply src/agent/onefuzz-agent/src/local/cmd.rs:91]: The default value is false for required.

File: src/agent/onefuzz-agent/src/tasks/fuzz/libfuzzer_fuzz.rs (hunk 5 above)
Comment (chkeita) [reply src/agent/onefuzz-agent/src/tasks/fuzz/libfuzzer_fuzz.rs:203]: good catch. this is a leftover of a debugging session. will remove

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+const BUFFER_SIZE: usize = 100;
+const TICK_RATE: Duration = Duration::from_millis(250);
+
+trait TakeAvailable<T> {
+    fn take_available(&mut self, size: usize) -> Result<Vec<T>>;
+}
+
+#[async_trait]
```
[the remark this answers is not in the document]
Comment (chkeita) [reply src/agent/onefuzz-agent/src/local/tui.rs:62]: that was the plan but the try_recv was not async. Will remove the attribute

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+impl<T: Send + Sync> TakeAvailable<T> for mpsc::UnboundedReceiver<T> {
+    fn take_available(&mut self, size: usize) -> Result<Vec<T>> {
+        let mut result = vec![];
+        while let Ok(v) = self.try_recv() {
```
[the remark this answers is not in the document]
Comment (chkeita) [reply src/agent/onefuzz-agent/src/local/tui.rs:66]: > why is this implemented with the synchronous `try_recv()`?
The async recv waits until data is available to return. I want this method to return an empty vector when no messages are available
> do we really want do the same thing for both an empty receiver and TryRecvError::Closed?
from the purpose of this method yes. a closed channel and an empty channel  should return an empty vector

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+
+    async fn refresh_ui(ui_state: UiLoopState) -> Result<UiLoopState, UiLoopError> {
+        let mut logs = ui_state.logs;
+        let mut file_count_state = ui_state.file_count_state;
+        let file_count = ui_state.file_count;
+        let mut log_event_receiver = ui_state.log_event_receiver;
+        let mut terminal = ui_state.terminal;
+
+        Self::take_available_logs(&mut log_event_receiver, 10, &mut logs);
+        //.extend_front(log_event_receiver.take_available(10)?.into_iter().rev());
```
Comment (ranweiler) [inline src/agent/onefuzz-agent/src/local/tui.rs:226]: Stray commented code

**Pull request merged**